### ISSUE TITLE
WINE issues

### ISSUE THREAD

**Issue by lojack5** [lojack5/IFileOperation#9]
It seems (some?, all?) WINE implementations only have stubs for IFileOperation.

WINE Info:
- Version: 8.1
- Related WINE issues:
  - https://bugs.winehq.org/show_bug.cgi?id=52350
  - https://bugs.winehq.org/show_bug.cgi?id=48848
- And related WINE code for IFileOperation implementation: https://github.com/wine-mirror/wine/blob/f9f4ee1b5882067b1900a67201c376f0a1abe3ea/dlls/shell32/shlfileop.c#L1803

Note that everything past the IUnknown interface methods are not implemented! So it seems IFileOperation is pretty much a no-go on WINE until they implement this.

In the meantime, we could probably catch this specific case and convert it to a more useful error (`NotImplementedError` is the natural one, but I'm thinking not to use that because it's not based on `OSError`, and could mask a intended `NotImplementedError` elsewhere in the code.  Then it's up to users of the library to decide how to proceed from there (use SHFileOperation instead, use native python methods, etc).

***
### Initial information:
Partial traceback:
```
 File "C:\Program Files\Mopy\bash\env\windows.py", line 109, in file_operation
    with FileOperator(parent, op_flgs) as fo:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\ifileoperation\fileoperator.py", line 137, in __init__
    self.sink_cookie = self.ifo.Advise(self.sink.com_ptr)

pywintypes.com_error: (-2147467263, 'Not implemented.', None, None)
```

And this part happens in a section not wrapped by `@convert_exceptions` since this was unexpected, so the correct looking HRESULT is `0x80004001` - E_NOTIMPL.

@llde on the Wrye Bash discord dug a little further and got:
```
0148:fixme:shell:file_operation_Advise (0000000000439D90, 000000000A7AD098, 0000000000209220): stub.
0148:fixme:shell:file_operation_QueryInterface not implemented for {df0b3d60-548f-101b-8e65-08002b2bd119}.
0148:fixme:shell:file_operation_Advise (00000000003AB430, 000000000A7AEC98, 0000000000209220): stub.
0148:fixme:shell:file_operation_QueryInterface not implemented for {df0b3d60-548f-101b-8e65-08002b2bd119}.
```
> Worst then ever all IFileOperation  is a stub in wine

So needs some more information:
- [ ] Host operating system information
- [x] WINE information (version, etc)
- [x] Any links to trackers for this issue for said WINE distributions (correct terminology?)

**Comment by llde** [comment]
wine is the last version 8.1
theis is the only bug opened about IFileOperation : https://bugs.winehq.org/show_bug.cgi?id=52350

**Comment by lojack5** [comment]
Thanks for the extra info!

**Comment by llde** [comment]
@lojack5 questions:
1) I see that the library set operational flags, add a moveFile op, set new flags and add a delete op. Do actions presrve the flags state (so that they are executed with that set of flags)?
2) What about 2 actions setted after a single set operation flags? 
3) Are actions executed in the order they are added? Or they are executed in bulk for type (all delete first for example)?
4) The sinks can be passed either with Advise or with the single operational functions. In case there are both, is one of them ignored or both are signalled? 
5) The sinks are singalled only in PerformOperations or also in the other functions?

**Comment by lojack5** [comment]
> @lojack5 questions:
> 
>     1. I see that the library set operational flags, add a moveFile op, set new flags and add a delete op. Do actions presrve the flags state (so that they are executed with that set of flags)?
> 
>     2. What about 2 actions setted after a single set operation flags?
> 
>     3. Are actions executed in the order they are added? Or they are executed in bulk for type (all delete first for example)?
> 
>     4. The sinks can be passed either with Advise or with the single operational functions. In case there are both, is one of them ignored or both are signalled?
> 
>     5. The sinks are singalled only in PerformOperations or also in the other functions?

1. Not sure your question here.  For Win32 IFileOperation, whatever flags its set with (via `SetOperationFlags`) are applicable to all operations queued with that instance.  Some flags have no meaning for certain operations (like FOF_RECYCLE doesn't have any effect on rename operations).  The way Wrye Bash uses `ifileoperation` causes a seperate `IFileOperation` instance to be made for each operation type.  **Edit:** Re-reading a few times, maybe your question is: What's the behavior of multiple calls to `SetOperationFlags` before a `PerformOperations` call?  Well, IFileOperation doesn't perform any file operations at all until `PerformOperations`, so whatever was set with the most recent `SetOperationFlags` is applied to a `PerformOperations`.
2. I think your question is: what if I use a single IFileOperation instance to do multiple operation types?  Like do a move, a copy, a rename, and a delete all at once?  See 1 above, the flags set with `SetOperationFlags` applies to all of them in that particular IFileOperation instance, but some flags have no effect on some operations.
3. No idea on this one.  I tried to look up on MSDN for more information on this, but no idea.  I *think* it's supposed to operate effectively atomically (since the whole operation can be undone if FOF_ALLOW_UNDO is specified), but obviously they aren't actually done simultaneously.
4. Not sure, but I think the notifications would be sent to both sinks, the docs aren't clear on that: https://learn.microsoft.com/en-us/windows/win32/api/shobjidl_core/nn-shobjidl_core-ifileoperation
5. Pretty sure it's only in PerformOperations.

**Comment by llde** [comment]
1 Yeah, I was unclear sorry. I wanted to ask wethever an operation would retain  the flags at the time of addition or if the flags would have depended on the last call of SetOperationFlags before PerformOperation.

**Comment by llde** [comment]
```
`
0174:fixme:shell:file_operation_SetOperationFlags (00000000003A15D0, 140614): stub- Returning SUCCESS.

0174:fixme:shell:file_operation_MoveItem (00000000003A15D0, 0000000000439C00, 000000000043EDC0, L"_tempfile.tmp", 0000000000000000): stub.

0174:fixme:shell:SupportErrorInfo_InterfaceSupportsErrorInfo (00000000003A15D8)->({947aab5f-0a5c-4c13-b4d6-4bf7836fc9f8})
0174:fixme:shell:file_operation_SetOperationFlags (00000000003A15D0, 210): stub- Returning SUCCESS.

0174:fixme:shell:file_operation_DeleteItem (00000000003A15D0, 0000000000439C00, 0000000000000000): stub.

0174:fixme:shell:file_operation_PerformOperations (00000000003A15D0): stub.
`

```
Here a Move operation and a DeleteOperation are being queued o the same IFIleOperation

**Comment by Utumno** [comment]
Another one https://bugs.winehq.org/show_bug.cgi?id=48848

**Comment by lojack5** [comment]
> Another one https://bugs.winehq.org/show_bug.cgi?id=48848

I love how in that one the last comment is "fixed" but all those patches actually did was add the stubs 🤣 

> Here a Move operation and a DeleteOperation are being queued o the same IFIleOperation

If that's coming from after running Wrye Bash I can see what's going on: we guard the early startup with some liberal `try: ... except Exception: ...` so unexpected exceptions don't completely kill it before we can get to a place to to be able to make a good enough debug log to help debug those exceptions (at least we try, doesn't always work).  So multiple IFileOperation operations are attempted before Wrye Bash completely quits, specifically:
- First a move is used to move a temp file into the target tested directory.
- Then that moved file is deleted, and the directory it came from (in separate calls)

So a "functioning IFileOperation" Wrye Bash startup which only has to test UAC should have the following calls done:
```
# Moving the temp file into target directory:
- Advise
- SetOperationFlags
- MoveFile
- PerformOperations
- GetAnyOperationsAborted
- Unadvise
# Deleting the temp file:
- Advise
- SetOperationFlags
- DeleteFile
- PerformOperations
- GetAnyOperationsAborted
- Unadvise
# Deleting the source temporary directory:
- Advise
- SetOperationFlags
- PerformOperations
- GetAnyOperationsAborted
- Unadvise
```

However, if any of those raise an exception (due to returning E_NOT_IMPLE), the whole sequence of calls gets interrupted and never finished.  Wrye Bash then gets rid of its IFileOperation reference (so pywin32 *should* be calling `IUknown:Release` on it to indicate it no longer holds a reference to it.

Now there's another thing going on in that log: the pointer to the IFileOperation instance: `00000000003A15D0` is the same in all calls. Tested on real Windows - this *only* is the same for the first two IFileOperation usages in Wrye Bash.  For testing I put in a bunch of garbage collection related code in python.  My only guess as to what's happening there is they aren't the same object (at least on real windows), just happen to occupy the same location in memory due to very close temporal proximity.  Don't know what WINE does (if it's deleting / creating a new one) or not.

**Comment by lojack5** [comment]
Also: I noticed some of those in your recent message are returning `SUCCESS`, is that due to local work on your end?

**Comment by llde** [comment]
Yup, I did implement Advise, Unadvise and Move/Delete functions. Now attepting to implement Move ops in PerformOperation.
HOnestly the documentation for this interface is horrible, not surprising very few applications seems to implement it to do shell file operations
